##edit mode and command mode

In [1]:
#cell for importing libraries
import pandas as pd
import numpy as np

In [2]:
#read the excel file
df=pd.read_excel("C:\\Users\\Yuvraj Singh Rathore\\ibmofferletter\\health_care_diabetes.xlsx")
df.head()


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

Pregnancies: This column indicates the number of times a woman has been pregnant.

Glucose: Referring to the blood sugar levels measured through a glucose test, this column holds data related to the glucose concentration in the blood.

BloodPressure: Representing the blood pressure measurement, this column contains values reflecting the pressure within the circulatory system.

SkinThickness: This column records the measurement of skin thickness, which might be a parameter related to health or body composition.

Insulin: Denoting the insulin level, this column contains data regarding the insulin concentration in the blood.

BMI (Body Mass Index): This column holds values calculated from a person's weight and height, serving as an indicator of body fatness and potential health risks associated with it.

DiabetesPedigreeFunction: This column contains a numerical value describing the likelihood of diabetes based on family history or genetic influence.

Age: Reflecting the age of the individuals in the dataset, this column records their ages at the time of data collection.

Outcome: This column denotes a binary outcome or label (commonly 0 or 1) indicating whether an individual shows signs of diabetes based on certain criteria or tests.

In [3]:
df.shape


(768, 9)

The output df.shape representing (768, 9) signifies that the DataFrame 'df' contains 768 rows and 9 columns. 

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [5]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

75th percentile in a dataset, it means that 75% of the data points fall below that particular value. So, if the 75th percentile of a dataset is, for example, 50, it indicates that 75% of the data points in that dataset are less than or equal to 50.

Pregnancies:

Most women in the dataset have had pregnancies, with an average of about 4.
However, some women have had none, while a few have had as many as 17 pregnancies, indicating a wide range of experiences.
Glucose:

The average glucose level is around 121 mg/dL, an important indicator in diabetes.
Some readings are as low as 0, potentially indicating missing or invalid data, while the highest recorded level is 199 mg/dL.
BloodPressure:

The average blood pressure is approximately 69 mm Hg.
There are records showing a blood pressure of 0, which might need further investigation, and the highest recorded pressure is 122 mm Hg.
SkinThickness:

On average, individuals have a skin thickness of around 21 units.
Some records show a skin thickness of 0, suggesting potential missing or incomplete data. The highest recorded thickness is 99 units.
Insulin:

The average insulin level is about 80 units.
There are instances of 0 insulin units recorded, which might indicate missing or incomplete data. The highest recorded level is 846 units.
BMI (Body Mass Index):

The average BMI is approximately 32, indicating a tendency toward higher body fat.
Some individuals have a BMI of 0, possibly indicating missing or inaccurate data. The highest recorded BMI is 67.1.
DiabetesPedigreeFunction:

The average diabetes pedigree function value is around 0.47, indicating a moderate genetic likelihood of diabetes.
The values range from 0.078 to 2.42, showcasing a varied genetic predisposition within the dataset.
Age:

The average age of individuals in the dataset is about 33 years.
The ages range from 21 to 81 years, demonstrating a broad spectrum of age groups.
Outcome:

Around 35% of individuals in the dataset exhibit signs of diabetes (value of 1).
The dataset is slightly skewed towards individuals without diabetes, as the 75th percentile indicates a value of 0 (no signs of diabetes) for 75% of the dataset.

### Data Preprocessing :Treating Missing Values"
#### In this dataset missing values represented by 0, so replace by  the mean of their feature columns.

In [6]:
#Finding number of having null values in a row
#df[i][df[i] != 0] =>[df[i] != 0] this is condtion return true or false and df[i] will take that values that has value true
for i in df.columns:
    print(i,(df[i]==0).sum())
           

Pregnancies 111
Glucose 5
BloodPressure 35
SkinThickness 227
Insulin 374
BMI 11
DiabetesPedigreeFunction 0
Age 0
Outcome 500


Pregnancies: It's plausible to have zero pregnancies. This doesn't necessarily require changing.
Glucose: Having zero glucose levels is improbable. so changing these values.
BloodPressure: Having zero blood pressure is unlikely and should be reviewed for correction.
SkinThickness: Zero skin thickness isn't realistic. These values may need to be modified.
Insulin: Zero insulin levels are improbable. we might need to address these values.
BMI: Having zero BMI is unrealistic. These values need attention.
DiabetesPedigreeFunction and Age: These columns have no rows with a value of zero, so they don't require any changes in this regard.
Outcome: Assuming this column represents diabetes (binary, possibly 0 for non-diabetic and 1 for diabetic), having 0 might be valid depending on the context.

now making list of these columns which have to be changed and replacing 0 by mean of the column

In [7]:
#making list of columns which contain null values or 0
change_col=[]
for i in df.columns:
    if (df[i]==0).sum()!=0:
        change_col.append(i)
#pregnancies,outcome can be 0 so removing these two
no_change=["Pregnancies","Outcome"]
for i in no_change:
    change_col.remove(i)
change_col

['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']

In [8]:
#finding the null value percentage 
null_percentage=(df[change_col]==0).mean()*100
null_percentage

Glucose           0.651042
BloodPressure     4.557292
SkinThickness    29.557292
Insulin          48.697917
BMI               1.432292
dtype: float64

(data[glucose]==0): This creates a boolean mask where each element in the 'glucose' column is checked for equality with 0. It will result in a series of True or False values.
(data[glucose]==0).mean(): This calculates the mean of the boolean values, treating True as 1 and False as 0. This effectively computes the proportion of occurrences where 'glucose' equals 0, which represents the proportion of True values in the boolean series.
* 100: Finally, multiplying by 100 converts this proportion to a percentage.
So, the expression (data[glucose]==0).mean()*100 would give you the percentage of rows in the 'glucose' column where the value is 0 in your dataset.

so glucose contain 0.6 percentage of rows as null values

Glucose: The 'Glucose' column contains approximately 65.10% null values.
BloodPressure: The 'BloodPressure' column contains approximately 4.56% null values.
SkinThickness: The 'SkinThickness' column contains approximately 29.56% null values.
Insulin: The 'Insulin' column contains approximately 48.70% null values.
BMI: The 'BMI' column contains approximately 1.43% null values.

In [9]:
#replacing 0 with the mean values of the columns
for i in change_col:
   df[i]=df[i].replace([0],df[i].mean())
#can also use interpolate method also 


Pregnancies  Glucose  BloodPressure  SkinThickness     Insulin   BMI  \
0              6    148.0           72.0      35.000000   79.799479  33.6   
1              1     85.0           66.0      29.000000   79.799479  26.6   
2              8    183.0           64.0      20.536458   79.799479  23.3   
3              1     89.0           66.0      23.000000   94.000000  28.1   
4              0    137.0           40.0      35.000000  168.000000  43.1   
..           ...      ...            ...            ...         ...   ...   
763           10    101.0           76.0      48.000000  180.000000  32.9   
764            2    122.0           70.0      27.000000   79.799479  36.8   
765            5    121.0           72.0      23.000000  112.000000  26.2   
766            1    126.0           60.0      20.536458   79.799479  30.1   
767            1     93.0           70.0      31.000000   79.799479  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [12]:

null_percentages = df.isnull().mean() * 100
null_percentages


Pregnancies                 0.0
Glucose                     0.0
BloodPressure               0.0
SkinThickness               0.0
Insulin                     0.0
BMI                         0.0
DiabetesPedigreeFunction    0.0
Age                         0.0
Outcome                     0.0
dtype: float64